### Business Decision:
A grocery store is interested in the buying trends of fruits and vegetables categories. It is planning to adjust the store layout based on the items purchased. 

Objective: To analyse the buying trends and provide appropriate suggestions.

In [2]:
# import library
import pandas as pd

In [3]:
# Load the data file
raw_df=pd.read_excel("D:\Business Analytics\Important files\Association mining\market_basket.xlsx")

#Examine the first few rows
raw_df.head(3)

Hair_Conditioner  Lemons  Standard_coffee  Frozen_Chicken_Wings  \
0                 0       0                0                     0   
1                 0       0                0                     0   
2                 0       0                0                     0   

   98pct_Fat_Free_Hamburger  Sugar_Cookies  Onions  Deli_Ham  \
0                         0              0       0         0   
1                         0              0       0         0   
2                         0              0       0         0   

   Dishwasher_Detergent  Beets  ...  Salt  Green_Beans  flav_Ice  \
0                     0      0  ...     0            0         0   
1                     0      0  ...     0            0         0   
2                     0      0  ...     0            0         0   

   Imported_Beer  Grits  Apple_Jelly  Beef_Jerky  Potatoes  Small_Eggs  \
0              0      0            0           0         0           0   
1              0      0            0           0         0           0   
2              0      0            0           0         0           0   

   Silver_Cleaner  
0               0  
1               0  
2               0  

[3 rows x 303 columns]

In [4]:
# Variable information
raw_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1361 entries, 0 to 1360
Columns: 303 entries, Hair_Conditioner to Silver_Cleaner
dtypes: int64(303)
memory usage: 3.1 MB


In [5]:
# Data summary
raw_df.describe()

Hair_Conditioner       Lemons  Standard_coffee  Frozen_Chicken_Wings  \
count        1361.00000  1361.000000      1361.000000           1361.000000   
mean            0.05878     0.030125         0.008817              0.026451   
std             0.23530     0.170994         0.093519              0.160532   
min             0.00000     0.000000         0.000000              0.000000   
25%             0.00000     0.000000         0.000000              0.000000   
50%             0.00000     0.000000         0.000000              0.000000   
75%             0.00000     0.000000         0.000000              0.000000   
max             1.00000     1.000000         1.000000              1.000000   

       98pct_Fat_Free_Hamburger  Sugar_Cookies       Onions     Deli_Ham  \
count               1361.000000    1361.000000  1361.000000  1361.000000   
mean                   0.093314       0.055107     0.080088     0.010287   
std                    0.290979       0.228272     0.271529     0.100937   
min                    0.000000       0.000000     0.000000     0.000000   
25%                    0.000000       0.000000     0.000000     0.000000   
50%                    0.000000       0.000000     0.000000     0.000000   
75%                    0.000000       0.000000     0.000000     0.000000   
max                    1.000000       1.000000     1.000000     1.000000   

       Dishwasher_Detergent        Beets  ...         Salt  Green_Beans  \
count            1361.00000  1361.000000  ...  1361.000000   1361.00000   
mean                0.02939     0.028655  ...     0.051433      0.01543   
std                 0.16896     0.166897  ...     0.220960      0.12330   
min                 0.00000     0.000000  ...     0.000000      0.00000   
25%                 0.00000     0.000000  ...     0.000000      0.00000   
50%                 0.00000     0.000000  ...     0.000000      0.00000   
75%                 0.00000     0.000000  ...     0.000000      0.00000   
max                 1.00000     1.000000  ...     1.000000      1.00000   

          flav_Ice  Imported_Beer        Grits  Apple_Jelly   Beef_Jerky  \
count  1361.000000    1361.000000  1361.000000  1361.000000  1361.000000   
mean      0.037472       0.024982     0.016899     0.033064     0.033064   
std       0.189986       0.156126     0.128942     0.178869     0.178869   
min       0.000000       0.000000     0.000000     0.000000     0.000000   
25%       0.000000       0.000000     0.000000     0.000000     0.000000   
50%       0.000000       0.000000     0.000000     0.000000     0.000000   
75%       0.000000       0.000000     0.000000     0.000000     0.000000   
max       1.000000       1.000000     1.000000     1.000000     1.000000   

          Potatoes   Small_Eggs  Silver_Cleaner  
count  1361.000000  1361.000000     1361.000000  
mean      0.086701     0.025716        0.007348  
std       0.281500     0.158346        0.085434  
min       0.000000     0.000000        0.000000  
25%       0.000000     0.000000        0.000000  
50%       0.000000     0.000000        0.000000  
75%       0.000000     0.000000        0.000000  
max       1.000000     1.000000        1.000000  

[8 rows x 303 columns]

In [114]:
#Creating a data frame with only fruits and vegetables.

selected_columns_df = raw_df[['Lemons', 'Onions', 'Beets', 'Sweat_Potatoes','Lettuce', 'Green_Pepper', 'Frozen_Broccoli', 'Canned_Peas', 'Whole_Green_Beans', 'Plums', 'Mushrooms', 'Dried_Apples', 'Carrots', 'Raisins', 'Frozen_Cauliflower', 'Tomatoes', 'Canned_Mixed_Fruit', 'Limes', 'Cabbage', 'Dried_Apricots', 'Peaches', 'Apples', 'Frozen_Carrots', 'Bananas', 'Pepper', 'Oranges', 'Cauliflower', 'Black-eyed_Peas', 'Mandarin_Oranges', 'Frozen_Peas', 'Garlic', 'Dried_Fruit_Mix', 'Broccoli', 'Mixed_Vegetables', 'Red_Pepper', 'Canned_Beets', 'Peas', 'Canned_Peaches', 'Sweet_Potatoes', 'Canned_Tomatoes', 'Green_Beans', 'Potatoes']] 

In [7]:
selected_columns_df.head(3)

Lemons  Onions  Beets  Sweat_Potatoes  Lettuce  Green_Pepper  \
0       0       0      0               0        0             0   
1       0       0      0               0        0             0   
2       0       0      0               0        0             0   

   Frozen_Broccoli  Canned_Peas  Whole_Green_Beans  Plums  ...  Broccoli  \
0                0            0                  0      0  ...         0   
1                0            0                  0      0  ...         0   
2                0            0                  0      1  ...         0   

   Mixed_Vegetables  Red_Pepper  Canned_Beets  Peas  Canned_Peaches  \
0                 0           0             0     0               0   
1                 0           0             0     0               0   
2                 0           0             0     0               0   

   Sweet_Potatoes  Canned_Tomatoes  Green_Beans  Potatoes  
0               0                0            0         0  
1               0                0            0         0  
2               0                0            0         0  

[3 rows x 42 columns]

In [57]:
import warnings
warnings.filterwarnings('ignore')

In [9]:
# import the apriori algorithm
from mlxtend.frequent_patterns import apriori

#### Identify frequent itemsets, i.e., items or item sets popular in the data sets.

##### Minimum support required in the data sets

In [12]:
# If a very high value of support is set, there won't be any combinations.

minimum_required_support=0.015

In [59]:
frequent_itemsets = apriori(selected_columns_df,min_support=minimum_required_support,use_colnames=True)

In [14]:
# The following items have support values greater than the minimum support models.

frequent_itemsets

support                      itemsets
0   0.030125                      (Lemons)
1   0.080088                      (Onions)
2   0.028655                       (Beets)
3   0.036003                     (Lettuce)
4   0.033064                (Green_Pepper)
..       ...                           ...
85  0.017634             (Garlic, Oranges)
86  0.022043           (Oranges, Potatoes)
87  0.026451            (Garlic, Potatoes)
88  0.016165  (Tomatoes, Onions, Potatoes)
89  0.015430   (Bananas, Onions, Potatoes)

[90 rows x 2 columns]

In [15]:
max_support_value = frequent_itemsets['support'].max()
max_support_value

0.0867009551800147

In [16]:
itemsets_with_max_support_value = frequent_itemsets[frequent_itemsets['support'] == max_support_value]['itemsets'].tolist()

print("Maximum Support:", max_support_value)
print("Itemsets with Maximum Support:", itemsets_with_max_support_value)

Maximum Support: 0.0867009551800147
Itemsets with Maximum Support: [frozenset({'Potatoes'})]


In [61]:
#Apply association mining algorithm using this support
frequent_itemsets= apriori(selected_columns_df, min_support=minimum_required_support, use_colnames=True)
frequent_itemsets

support                      itemsets
0   0.030125                      (Lemons)
1   0.080088                      (Onions)
2   0.028655                       (Beets)
3   0.036003                     (Lettuce)
4   0.033064                (Green_Pepper)
..       ...                           ...
85  0.017634             (Garlic, Oranges)
86  0.022043           (Oranges, Potatoes)
87  0.026451            (Garlic, Potatoes)
88  0.016165  (Tomatoes, Onions, Potatoes)
89  0.015430   (Bananas, Onions, Potatoes)

[90 rows x 2 columns]

In [18]:
from mlxtend.frequent_patterns import association_rules

In [19]:
## Set rules from the frequent itemsets that match the desired lift.

## Set the required lift

minimum_required_lift = 1.1
rules = association_rules(frequent_itemsets,metric="lift", min_threshold=minimum_required_lift)

In [20]:
rules

antecedents          consequents  antecedent support  \
0               (Onions)             (Lemons)            0.080088   
1               (Lemons)             (Onions)            0.030125   
2                (Plums)             (Onions)            0.055841   
3               (Onions)              (Plums)            0.080088   
4               (Onions)          (Mushrooms)            0.080088   
..                   ...                  ...                 ...   
103  (Bananas, Potatoes)             (Onions)            0.024982   
104   (Onions, Potatoes)            (Bananas)            0.033064   
105            (Bananas)   (Onions, Potatoes)            0.059515   
106             (Onions)  (Bananas, Potatoes)            0.080088   
107           (Potatoes)    (Bananas, Onions)            0.086701   

     consequent support   support  confidence      lift  leverage  conviction  \
0              0.030125  0.016899    0.211009  7.004475  0.014487    1.229260   
1              0.080088  0.016899    0.560976  7.004475  0.014487    2.095355   
2              0.080088  0.022777    0.407895  5.093071  0.018305    1.553629   
3              0.055841  0.022777    0.284404  5.093071  0.018305    1.319401   
4              0.054372  0.020573    0.256881  4.724523  0.016219    1.272512   
..                  ...       ...         ...       ...       ...         ...   
103            0.080088  0.015430    0.617647  7.712089  0.013429    2.405923   
104            0.059515  0.015430    0.466667  7.841152  0.013462    1.763409   
105            0.033064  0.015430    0.259259  7.841152  0.013462    1.305364   
106            0.024982  0.015430    0.192661  7.712089  0.013429    1.207693   
107            0.023512  0.015430    0.177966  7.569121  0.013391    1.187892   

     zhangs_metric  
0         0.931866  
1         0.883860  
2         0.851186  
3         0.873622  
4         0.856972  
..             ...  
103       0.892633  
104       0.902301  
105       0.927679  
106       0.946105  
107       0.950274  

[108 rows x 10 columns]

In [65]:
print("Total number of rules generated by the algorithm with given support and lift:",len(rules))

Total number of rules generated by the algorithm with given support and lift: 108


In [95]:
# Set the minimum threshold for confidence
minimum_required_confidence = 0.5

In [97]:
# Filter rules based on confidence and lift threshold
filtered_rules = rules[(rules['confidence'] >= minimum_required_confidence)]

In [99]:
filtered_rules

antecedents consequents  antecedent support  consequent support  \
1                (Lemons)    (Onions)            0.030125            0.080088   
21       (Canned_Peaches)    (Onions)            0.025716            0.080088   
96     (Tomatoes, Onions)  (Potatoes)            0.028655            0.086701   
97   (Tomatoes, Potatoes)    (Onions)            0.028655            0.080088   
102     (Bananas, Onions)  (Potatoes)            0.023512            0.086701   
103   (Bananas, Potatoes)    (Onions)            0.024982            0.080088   

      support  confidence      lift  leverage  conviction  zhangs_metric  
1    0.016899    0.560976  7.004475  0.014487    2.095355       0.883860  
21   0.016165    0.628571  7.848493  0.014105    2.476686       0.895619  
96   0.016165    0.564103  6.506302  0.013680    2.095215       0.871269  
97   0.016165    0.564103  7.043519  0.013870    2.110386       0.883338  
102  0.015430    0.656250  7.569121  0.013391    2.656870       0.888781  
103  0.015430    0.617647  7.712089  0.013429    2.405923       0.892633

In [101]:
print("Total number of filtered rules to meet the lift, confidence and support criteria:",len(filtered_rules))

Total number of filtered rules to meet the lift, confidence and support criteria: 6


In [103]:
# Sorting the rules based on lift, confidence, and support (all in descending order)

sorted_rules = filtered_rules.sort_values(by=['lift','confidence','support'],ascending=False)

In [105]:
sorted_rules

antecedents consequents  antecedent support  consequent support  \
21       (Canned_Peaches)    (Onions)            0.025716            0.080088   
103   (Bananas, Potatoes)    (Onions)            0.024982            0.080088   
102     (Bananas, Onions)  (Potatoes)            0.023512            0.086701   
97   (Tomatoes, Potatoes)    (Onions)            0.028655            0.080088   
1                (Lemons)    (Onions)            0.030125            0.080088   
96     (Tomatoes, Onions)  (Potatoes)            0.028655            0.086701   

      support  confidence      lift  leverage  conviction  zhangs_metric  
21   0.016165    0.628571  7.848493  0.014105    2.476686       0.895619  
103  0.015430    0.617647  7.712089  0.013429    2.405923       0.892633  
102  0.015430    0.656250  7.569121  0.013391    2.656870       0.888781  
97   0.016165    0.564103  7.043519  0.013870    2.110386       0.883338  
1    0.016899    0.560976  7.004475  0.014487    2.095355       0.883860  
96   0.016165    0.564103  6.506302  0.013680    2.095215       0.871269

In [107]:
# Selecting the top 5 rules

top_5_rules = sorted_rules.head(5)

In [109]:
# Print only the support, confidence, and lift column

top_5_rules[['antecedents','consequents', 'support', 'confidence', 'lift']]

antecedents consequents   support  confidence      lift
21       (Canned_Peaches)    (Onions)  0.016165    0.628571  7.848493
103   (Bananas, Potatoes)    (Onions)  0.015430    0.617647  7.712089
102     (Bananas, Onions)  (Potatoes)  0.015430    0.656250  7.569121
97   (Tomatoes, Potatoes)    (Onions)  0.016165    0.564103  7.043519
1                (Lemons)    (Onions)  0.016899    0.560976  7.004475

### Inference:

Lift of frequently bought item no 21: Canned_Peaches is greater than the item no 103. 
People who buy canned-peaches also buy onions.
Simialrly,  customers who buy bananas and potatoes in this order also buy onions.

### Business decision:

The grocery store can adjust its layout so that onions are kept next to or close to where canned_peaches are stored. Similarly. it can keep onions, next to bananas and potatoes.